File: src/main/java/net/fellter/moreLeafParticles/ModParticles.java
//~ not_null
//~ non_null_import

package net.fellter.moreLeafParticles;

import org.jetbrains.annotations.NotNull;

import net.minecraft.core.Registry;
import net.minecraft.core.particles.ColorParticleOption;
import net.minecraft.core.particles.ParticleType;
import net.minecraft.core.particles.SimpleParticleType;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;

import net.fabricmc.fabric.api.particle.v1.FabricParticleTypes;

public final class ModParticles {
	public static final ParticleType<@NotNull ColorParticleOption> SPRUCE_NEEDLES = register("spruce_needles");
	public static final ParticleType<@NotNull ColorParticleOption> BIRCH_LEAVES = register("birch_leaves");
	public static final ParticleType<@NotNull ColorParticleOption> MANGROVE_LEAVES = register("mangrove_leaves");
	public static final ParticleType<@NotNull ColorParticleOption> JUNGLE_LEAVES = register("jungle_leaves");
	public static final ParticleType<@NotNull ColorParticleOption> ACACIA_LEAVES = register("acacia_leaves");
	public static final ParticleType<@NotNull ColorParticleOption> DARK_OAK_LEAVES = register("dark_oak_leaves");
	public static final SimpleParticleType AZALEA_LEAVES = registerSimple("azalea_leaves");
	public static final SimpleParticleType FLOWERING_AZALEA_PARTICLES = registerSimple("flowering_azalea_particles");

	private static ParticleType<@NotNull ColorParticleOption> register(String name) {
		ParticleType<@NotNull ColorParticleOption> type = FabricParticleTypes.complex(ColorParticleOption::codec, ColorParticleOption::streamCodec);
		Registry.register(BuiltInRegistries.PARTICLE_TYPE, ResourceLocation.fromNamespaceAndPath(MoreLeafParticles.MOD_ID, name), type);
		return type;
	}

	private static SimpleParticleType registerSimple(String name) {
		SimpleParticleType type = FabricParticleTypes.simple(false);
		Registry.register(BuiltInRegistries.PARTICLE_TYPE, ResourceLocation.fromNamespaceAndPath(MoreLeafParticles.MOD_ID, name), type);
		return type;
	}

	public static void init() {
	}
}

File: src/main/java/net/fellter/moreLeafParticles/ModParticles.kt
package net.fellter.moreLeafParticles

import net.fabricmc.fabric.api.particle.v1.FabricParticleTypes
import net.minecraft.core.Registry
import net.minecraft.core.particles.ColorParticleOption
import net.minecraft.core.particles.ParticleType
import net.minecraft.core.particles.SimpleParticleType
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.Identifier

object ModParticles {
	@JvmField val SPRUCE_NEEDLES: ParticleType<ColorParticleOption> = register("spruce_needles")
	@JvmField val BIRCH_LEAVES: ParticleType<ColorParticleOption> = register("birch_leaves")
	@JvmField val MANGROVE_LEAVES: ParticleType<ColorParticleOption> = register("mangrove_leaves")
	@JvmField val JUNGLE_LEAVES: ParticleType<ColorParticleOption> = register("jungle_leaves")
	@JvmField val ACACIA_LEAVES: ParticleType<ColorParticleOption> = register("acacia_leaves")
	@JvmField val DARK_OAK_LEAVES: ParticleType<ColorParticleOption> = register("dark_oak_leaves")
	@JvmField val AZALEA_LEAVES: SimpleParticleType = registerSimple("azalea_leaves")
	@JvmField val FLOWERING_AZALEA_PARTICLES: SimpleParticleType = registerSimple("flowering_azalea_particles")

	private fun register(name: String): ParticleType<ColorParticleOption> {
		return Registry.register(
			BuiltInRegistries.PARTICLE_TYPE, Identifier.fromNamespaceAndPath(MoreLeafParticles.MOD_ID, name),
			FabricParticleTypes.complex(ColorParticleOption::codec, ColorParticleOption::streamCodec)
		)
	}

	private fun registerSimple(name: String): SimpleParticleType {
		return Registry.register(
			BuiltInRegistries.PARTICLE_TYPE, Identifier.fromNamespaceAndPath(MoreLeafParticles.MOD_ID, name),
			FabricParticleTypes.simple(false)
		)
	}

	@JvmStatic fun init() {}
}
